### PR TITLE
Update implementation of Basic02

### LINKED ISSUE #1193: Update implementation of updated BASIC02
The specification of BASIC02 has been updated by https://github.com/zonemaster/zonemaster/pull/1085. The implementation must be updated to match the specification.

### PR DESCRIPTION
**Pull request by tgreenx**
## Purpose

This PR proposes an update implementation of Basic02 to latest specification (https://github.com/zonemaster/zonemaster/pull/1085 and https://github.com/zonemaster/zonemaster/pull/1145).

## Context

Fixes #1193, #881 

## Changes

Update implementation of Basic02:
- Logic
- Message tags

Update unitary tests:

- Some are entirely new
- Modernized utility functions

## How to test this PR

Unitary tests should pass.

Also (without an IPv6 connectivity):
```
$ zonemaster-cli --test basic/basic02 zonemaster.net --level INFO --show-testcase --raw --no-ipv6
   0.00 INFO      UNSPECIFIED    GLOBAL_VERSION   version=v4.6.2
   6.09 INFO      BASIC02        B02_AUTH_RESPONSE_SOA   domain="zonemaster.net"; ns_list=ns2.nic.fr/192.93.0.4;nsa.dnsnode.net/194.58.192.46;nsp.dnsnode.net/194.58.198.32;nsu.dnsnode.net/185.42.137.98
```

And (from #881):
```
$ zonemaster-cli --test basic/basic02 e.consistency05.exempelvis.se --level INFO --show-testcase --raw --no-ipv6
   0.00 INFO      UNSPECIFIED    GLOBAL_VERSION   version=v4.6.2
   1.37 CRITICAL  BASIC02        B02_NO_WORKING_NS   domain="e.consistency05.exempelvis.se"
   1.37 ERROR     BASIC02        B02_UNEXPECTED_RCODE   ns=ns2.e.consistency05.exempelvis.se/194.18.226.122; rcode=NXDOMAIN
   1.37 ERROR     BASIC02        B02_UNEXPECTED_RCODE   ns=ns1.e.consistency05.exempelvis.se/46.21.97.97; rcode=REFUSED
```

### CODE REVIEW

**Comment by matsduf** [comment]
```
t/util.t .................. 1/? unable to parse RR string at /usr/local/lib/perl5/site_perl/Net/DNS/ZoneFile.pm line 534.
 at /usr/home/matsd/.cpanm/work/1677164454.51730/Zonemaster-Engine-v4.6.1/blib/lib/Zonemaster/Engine/Util.pm line 179.
```

**Comment by matsduf** [comment]
I tested the name lame.dufberg.se that is a lame delegation. The expected message is B02_NO_WORKING_NS and B02_NS_NO_RESPONSE but that is not what we get.

```
# zonemaster-cli --test basic/basic02 lame.dufberg.se --level INFO --show-testcase --raw
   0.00 INFO      UNSPECIFIED    GLOBAL_VERSION   version=v4.6.1
  20.86 CRITICAL  BASIC02        B02_NO_DELEGATION   domain="lame.dufberg.se"
```

**Comment by matsduf** [comment]
> This error is also present on master and develop branches, actually.

You are right. That is a false error message. It should be hidden.

File: lib/Zonemaster/Engine/Test/Basic.pm [hunk 1]
```diff
@@ -381,33 +414,104 @@ sub basic02 {
         }
         _ip_enabled_message( \@results, $ns, $query_type );
 
+        # This is not a realistical conditional check considering the current implementation of Engine::Nameserver.
+        # Any Engine::Nameserver object created will (must) have an IP address. So it is here as a placeholder.
+        if ( not $ns->address ) {
+            $ns_cant_resolve{$ns->name} = 1;
+            next;
+        }
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/Basic.pm:422]: That limitation is really bad. That means that we cannot see when a domain is delegated to a name server name that cannot be resolved into an address.

File: lib/Zonemaster/Engine/Test/Basic.pm [hunk 2]
```diff
@@ -365,14 +385,27 @@ sub basic01 {
...
+    my $query_type = q{SOA};
+
+    my %auth_response_soa;
+    my %ns_broken;
+    my %ns_not_auth;
+    my %ns_cant_resolve;
+    my %ns_no_response;
+    my %unexpected_rcode;
+
     my @ns = @{ Zonemaster::Engine::TestMethods->method4( $zone ) };
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/Basic.pm:398]: Does metod4 return an empty array when the delegation is completely lame?

**Review by matsduf — CHANGES_REQUESTED** [review]

File: lib/Zonemaster/Engine/Test/Basic.pm (hunk 2 above)
Comment (pnax) [reply lib/Zonemaster/Engine/Test/Basic.pm:398]: > does the implemented Method4 include non-glue address records

Digging into the implementation, I would answer yes (however, I've not tested it yet):
https://github.com/zonemaster/zonemaster-engine/blob/0fdd78026307e306793e4415b60d8c490e670110/lib/Zonemaster/Engine/TestMethods.pm#L40-L44
https://github.com/zonemaster/zonemaster-engine/blob/0fdd78026307e306793e4415b60d8c490e670110/lib/Zonemaster/Engine/Zone.pm#L56-L77
https://github.com/zonemaster/zonemaster-engine/blob/0fdd78026307e306793e4415b60d8c490e670110/lib/Zonemaster/Engine/NSArray.pm#L137-L152
ref: https://perldoc.perl.org/perltie (I'm discovering this tie function, never used nor tested it yet.)

Comment (matsduf) [reply lib/Zonemaster/Engine/Test/Basic.pm:398]: > > does the implemented Method4 include non-glue address records
> 
> Digging into the implementation, I would answer yes (however, I've not tested it yet):

Well, then it is probably enough if the implementation uses Method2 to get the NS in the delegation and Method4 to get nameserver IP addresses using Method4.

**Comment by matsduf** [comment]
> @matsduf @pnax I think I got to a suitable workaround for both issues. See commit [9e700c7](https://github.com/zonemaster/zonemaster-engine/commit/9e700c702327d2ee96298fe828a4bcb1bef64ca3). Let me know what you think (note that unitary tests data are still not updated and thus will fail).

I am not sure that it is a work-around or just the solution for the methods that are available, but it looks fine. I will create some more test delegations for related scenarios to see if they behave the same.

**Comment by matsduf** [comment]
I created yet another test zone (or rather test delegation):
```
zonemaster-cli --show-testcase --test basic/basic02 lame-ns-servfail.dufberg.se. --level INFO --raw
   0.00 INFO      UNSPECIFIED    GLOBAL_VERSION   version=v4.6.1
   0.55 CRITICAL  BASIC02        B02_NO_WORKING_NS   domain="lame-ns-servfail.dufberg.se"
   0.55 ERROR     BASIC02        B02_NS_NOT_AUTH   ns=dns3.narnia.pp.se/2a05:d014:4ae:e900:e026:2d09:8d68:2d43
   0.55 ERROR     BASIC02        B02_NS_NOT_AUTH   ns=dns3.narnia.pp.se/3.124.111.178
```

> Good to hear! If you want, I can make a quick update to the specification and swap those two checks around, so that the implementation can be updated too. It seems like a very small thing to do.

Please review https://github.com/zonemaster/zonemaster/pull/1145

> Also, do you want me to add those zones to the unitary tests?

Yes, that would be good. 

After an update based on https://github.com/zonemaster/zonemaster/pull/1145, the following zones should return B02_NS_NO_IP_ADDR:
* lame-ns-no-name.dufberg.se.
* lame-ns-no-addr.dufberg.se.
* lame-ns-no-glue.dufberg.se.

The following B02_NS_NO_RESPONSE:
* lame-ns-no-response.dufberg.se.

The following B02_UNEXPECTED_RCODE:
* lame-ns-refused.dufberg.se.
* lame-ns-servfail.dufberg.se.

**Comment by matsduf** [comment]
Here is another test:

```
zonemaster-cli --show-testcase --test basic/basic02  --level INFO --raw www.zonemaster.net --ns nsa.dnsnode.net
   0.00 INFO      UNSPECIFIED    GLOBAL_VERSION   version=v4.6.2
   0.11 CRITICAL  BASIC02        B02_NO_WORKING_NS   domain="www.zonemaster.net"
   0.11 ERROR     BASIC02        B02_NS_BROKEN   ns=nsa.dnsnode.net/2a01:3f1:46::53
   0.11 ERROR     BASIC02        B02_NS_BROKEN   ns=nsa.dnsnode.net/194.58.192.46
```

**Comment by matsduf** [comment]
This delegation expects B02_NS_NOT_AUTH:
```
$ zonemaster-cli --show-testcase --test basic/basic02  --level INFO --raw zonemaster.net --ns g.root-servers.net --ns a.root-servers.net
   0.00 INFO      UNSPECIFIED    GLOBAL_VERSION   version=v4.6.2
   0.26 CRITICAL  BASIC02        B02_NO_WORKING_NS   domain="zonemaster.net"
   0.26 ERROR     BASIC02        B02_NS_NOT_AUTH   ns=g.root-servers.net/2001:500:12::d0d
   0.26 ERROR     BASIC02        B02_NS_NOT_AUTH   ns=a.root-servers.net/2001:503:ba3e::2:30
   0.26 ERROR     BASIC02        B02_NS_NOT_AUTH   ns=a.root-servers.net/198.41.0.4
   0.26 ERROR     BASIC02        B02_NS_NOT_AUTH   ns=g.root-servers.net/192.112.36.4
```

**Review by matsduf — APPROVED** [review]

**Review by matsduf — APPROVED** [review]

**Comment by matsduf** [comment]
> Now we have full testing coverage for Basic02. Please re-review.

That is really great.

**Pull request merged**